chore(helm): Improve health probe helper templates

lindhe opened the pull request:
**What this PR does / why we need it**:

The logic for the health probes' helper templates were pretty complex. This change rewrites the template definitions so they are easier to read and understand.

**Which issue(s) this PR fixes**:

None.

**Special notes for your reviewer**:

Replaces part of #15359

**Checklist**

- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [ ] Tests updated
- [x] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

jkroepke commented:
> There's an extra {{- end -}} with not apparent start, yet it did not render an error in the templates (which is very surprising).

There is an with-else and within the else block, there is an with again. You merge that into else if condition which requires one less end block.

But this this change only cosmetic?

lindhe commented:
> > There's an extra {{- end -}} with not apparent start, yet it did not render an error in the templates (which is very surprising).
> 
> There is an with-else and within the else block, there is an with again. You merge that into else if condition which requires one less end block.
> 
> But this this change only cosmetic?

Oh. I was so confused by the extra `{{- else }}` clause and hard-to-follow logic with the nested `with` statements, so I miscounted the number of required `{{ end }}` clauses. You are right, the fact that I have one less `{{ end }}` now is irrelevant. I'll remove that comment from my PR description.

And just to convince myself that the old version was indeed invalid, here's a commented example from the old version:

```text
{{/*
bloom gateway readinessProbe
*/}}
{{- define "loki.bloomGateway.readinessProbe" -}} 🟢 Here `define` begins.
{{- with .Values.bloomGateway.readinessProbe }} 🟡 Here the outer `with` begins.
readinessProbe:
  {{- toYaml . | nindent 2 }}
{{- else }} ❌ This `else` clause seems invalid, since there's no `if` statement around here.
{{- with .Values.loki.readinessProbe }} 🔴 Here the inner `with` begins.
readinessProbe:
  {{- toYaml . | nindent 2 }}
{{- end }} {{/* 🔴 Here the inner `with` ends. */}}
{{- end }} {{/* 🟡 Here the outer `with` ends. */}}
{{- end -}}  {{/* 🟢 Here `define` ends */}}
```

I think it's quite clear that the `{{- else }}` clause is dangling and should be removed. After removing the invalid `{{- else }}`, we have a `with` nested inside another `with`:

1. If `.Values.bloomGateway.readinessProbe` is false, then the outer `with` statement will not render, thus neither `readinessProbe` gets rendered. This feel wrong.
1. If `.Values.bloomGateway.readinessProbe` is true and  `.Values.loki.readinessProbe` is false, then the rendered output is simple enough:

    ```yaml
    readinessProbe:
      # probes from .Values.bloomGateway.readinessProbe
    ```

1. And if both `.Values.bloomGateway.readinessProbe` and `.Values.loki.readinessProbe` are true, then I think the rendered content would become:

    ```yaml
    readinessProbe:
      # probes from .Values.bloomGateway.readinessProbe
    readinessProbe:
      # probes from .Values.loki.readinessProbe
    ```

    My experience with duplicate keys in Kubernetes is that only the last key gets rendered, which in this case would mean that `.Values.loki.readinessProbe` takes precedence over `.Values.bloomGateway.readinessProbe` in the `bloom-gateway` component. I do not think that makes sense. Also, as far as I can tell, duplicate keys is invalid according to the YAML spec and is undefined behavior in Kubernetes' YAML parser, so it's probably not good to rely on that behavior.

I believe my version addresses these issues.

jkroepke commented:
So the issue to solve here is duplicate keys in the templates and simply the logic, right? The else you mention is not invalid. The with-else-end is documented at go. Ref: https://pkg.go.dev/text/template

```
{{with pipeline}} T1 {{else}} T0 {{end}}
	If the value of the pipeline is empty, dot is unaffected and T0
	is executed; otherwise, dot is set to the value of the pipeline
	and T1 is executed.
```

But I agree, no one understand that.

lindhe commented:
> The with-else-end is documented at go. Ref: https://pkg.go.dev/text/template
> 
> ```
> {{with pipeline}} T1 {{else}} T0 {{end}}
> 	If the value of the pipeline is empty, dot is unaffected and T0
> 	is executed; otherwise, dot is set to the value of the pipeline
> 	and T1 is executed.
> ```

Oh. I was pouring over Helm's template documentation and it was nowhere to be found. If it's valid Go template, then I'm guessing it's valid in Helm templates too, just that it's undocumented.

lindhe commented on production/helm/loki/CHANGELOG.md:
```diff
@@ -16,6 +16,7 @@ Entries should include a reference to the pull request that introduced the chang
 - [FEATURE] Added readiness probe for memcached [#15609](https://github.com/grafana/loki/pull/15609)
 - [FEATURE] Added pdb for pattern ingester [#17058](https://github.com/grafana/loki/pull/17058)
 - [BUGFIX] Ensure ui.enabled=true is set in loki ConfigMap when loki.ui.enabled=true is set in values.yaml to actually enable the UI [#17562](https://github.com/grafana/loki/pull/17562)
+- [BUGFIX] Fix broken `readinessProbe` and `livenessProbe` helper template definitons. [#18347](https://github.com/grafana/loki/pull/18347)
```
If `with` in Helm templates indeed has `else` clauses like in Go templates as you pointed out, then I guess this should be reworded from "Fix broken" to "Streamline" or something like that. WDYT?

lindhe commented:
> If feel your PR still valid and worthy, if it simply the logic.

If you want it, then I'll reopen it! 😄 Will check your comments soon.

lindhe commented on production/helm/loki/templates/bloom-gateway/_helpers-bloom-gateway.tpl:
```diff
@@ -24,17 +24,15 @@ app.kubernetes.io/component: bloom-gateway
 readinessProbe:
-  {{- toYaml . | nindent 2 }}
-{{- else }}
-{{- with .Values.loki.readinessProbe }}
+  {{- toYaml .Values.bloomGateway.readinessProbe | nindent 2 }}
+{{- else if .Values.loki.readinessProbe }}
 readinessProbe:
-  {{- toYaml . | nindent 2 }}
+  {{- toYaml .Values.loki.readinessProbe | nindent 2 }}
 {{- end }}
```
Very elegant, nice one! 👌 I'll implement this throughout the PR.

lindhe commented:
@jkroepke Are we waiting for anything else before approval?

Jayclifford345 reviewed (approved):
LGTM!
